File: chapter06/mrtr-client/python/input_handler.py
# handle_input_request: the client host's side of an elicitation.
#
# Two concerns from Chapter 6, Section 6.2.5:
#   1. Form rendering — a schema-driven console form with labels, help text,
#      defaults, the accept/decline/cancel three-action model, and validation
#      of the user's input against the requested schema before responding.
#   2. Policy hooks — pre-user policies that can auto-answer or deny a
#      request without it ever reaching the user.
#
# Modes:
#   - Interactive (default): renders the form on the terminal.
#   - Scripted: MRTR_ANSWERS holds a JSON array of predetermined answers,
#     consumed in order — one per elicitation. An entry is either a content
#     object (meaning accept) or { "action": "decline" | "cancel" }.
#     Example: MRTR_ANSWERS='[{"title":"Sync","duration":"30"},{"confirm":true}]'
#
# MRTR_POLICY=autoconfirm enables the demo auto-answer policy.

import json
import os
import re
from typing import Any

import anyio
from mcp_types import (
    ElicitRequestFormParams,
    ElicitRequestParams,
    ElicitRequestURLParams,
    ElicitResult,
    InputRequest,
)

# ---------------------------------------------------------------------------
# Policy hooks: run before the user ever sees the request. A policy returns a
# full response to short-circuit the form, or None to pass.
# ---------------------------------------------------------------------------

_SUSPICIOUS = re.compile(r"password|api[ -]?key|token|secret|credential", re.IGNORECASE)


def deny_credential_solicitation(params: ElicitRequestFormParams) -> ElicitResult | None:
    """Form mode must never collect credentials; deny requests that look like it."""
    for key, prop in (params.requested_schema.get("properties") or {}).items():
        texts = " ".join([key, prop.get("title") or "", prop.get("description") or ""])
        if prop.get("format") == "password" or _SUSPICIOUS.search(texts):
            print(f'  [policy] declined: field "{key}" appears to solicit credentials')
            return ElicitResult(action="decline")
    return None


def auto_confirm(params: ElicitRequestFormParams) -> ElicitResult | None:
    """Demo auto-answer policy: accept pure-confirmation forms (all-boolean) with their defaults."""
    if os.environ.get("MRTR_POLICY") != "autoconfirm":
        return None
    props = (params.requested_schema.get("properties") or {}).items()
    if not props or not all(p.get("type") == "boolean" for _, p in props):
        return None
    content = {key: prop.get("default") if isinstance(prop.get("default"), bool) else True for key, prop in props}
    print(f"  [policy] auto-answered confirmation form: {json.dumps(content)}")
    return ElicitResult(action="accept", content=content)


POLICIES = [deny_credential_solicitation, auto_confirm]

# ---------------------------------------------------------------------------
# Validation: check content against the requested schema before responding —
# a correctness concern (the server expects this shape) and a security one.
# ---------------------------------------------------------------------------


def validate_field(key: str, prop: dict[str, Any], value: Any) -> str | None:
    if (options := prop.get("enum")) is not None:
        if not isinstance(value, str) or value not in options:
            return f'"{key}" must be one of: {", ".join(options)}'
        return None
    match prop.get("type"):
        case "string":
            if not isinstance(value, str):
                return f'"{key}" must be a string'
            if (n := prop.get("minLength")) is not None and len(value) < n:
                return f'"{key}" must be at least {n} characters'
            if (n := prop.get("maxLength")) is not None and len(value) > n:
                return f'"{key}" must be at most {n} characters'
            return None
        case "number" | "integer" as t:
            if isinstance(value, bool) or not isinstance(value, (int, float)):
                return f'"{key}" must be a number'
            if t == "integer" and not float(value).is_integer():
                return f'"{key}" must be an integer'
            if (n := prop.get("minimum")) is not None and value < n:
                return f'"{key}" must be >= {n}'
            if (n := prop.get("maximum")) is not None and value > n:
                return f'"{key}" must be <= {n}'
            return None
        case "boolean":
            return None if isinstance(value, bool) else f'"{key}" must be a boolean'
        case t:
            return f'"{key}" has an unsupported schema type: {t}'


def validate_content(params: ElicitRequestFormParams, content: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    properties = params.requested_schema.get("properties") or {}
    for key in params.requested_schema.get("required") or []:
        if key not in content:
            errors.append(f'"{key}" is required')
    for key, value in content.items():
        prop = properties.get(key)
        if prop is None:
            errors.append(f'"{key}" is not in the requested schema')
            continue
        if (error := validate_field(key, prop, value)) is not None:
            errors.append(error)
    return errors


def defaults_of(params: ElicitRequestFormParams) -> dict[str, Any]:
    """Pre-populate defaults, per the schema-driven rendering rules."""
    properties = params.requested_schema.get("properties") or {}
    return {key: prop["default"] for key, prop in properties.items() if "default" in prop}


# ---------------------------------------------------------------------------
# Scripted answers (for smoke tests / headless runs).
# ---------------------------------------------------------------------------

_scripted_answers: list[Any] | None = (
    json.loads(os.environ["MRTR_ANSWERS"]) if os.environ.get("MRTR_ANSWERS") else None
)


def next_scripted_answer(params: ElicitRequestFormParams) -> ElicitResult:
    assert _scripted_answers is not None
    if not _scripted_answers:
        raise RuntimeError("MRTR_ANSWERS ran out of scripted answers")
    entry = _scripted_answers.pop(0)
    if entry.get("action") in ("decline", "cancel"):
        print(f"  [scripted] {entry['action']}")
        return ElicitResult(action=entry["action"])
    content = {**defaults_of(params), **entry}
    if errors := validate_content(params, content):
        raise RuntimeError(f"Scripted answer failed validation: {'; '.join(errors)}")
    print(f"  [scripted] accept: {json.dumps(content)}")
    return ElicitResult(action="accept", content=content)


# ---------------------------------------------------------------------------
# Interactive form rendering.
# ---------------------------------------------------------------------------


def _parse_input(prop: dict[str, Any], raw: str) -> tuple[bool, Any]:
    if prop.get("enum") is not None or prop.get("type") == "string":
        return True, raw
    if prop.get("type") in ("number", "integer"):
        try:
            return True, int(raw) if prop["type"] == "integer" else float(raw)
        except ValueError:
            return False, None
    if prop.get("type") == "boolean":
        if raw.lower() in ("y", "yes", "true"):
            return True, True
        if raw.lower() in ("n", "no", "false"):
            return True, False
    return False, None


def _render_form_sync(params: ElicitRequestFormParams) -> ElicitResult:
    print(f"\n{params.message}")
    print("(enter a value, or !decline / !cancel; blank accepts the default)")
    content: dict[str, Any] = {}
    required = params.requested_schema.get("required") or []
    for key, prop in (params.requested_schema.get("properties") or {}).items():
        label = prop.get("title") or key  # title is the display label
        if prop.get("description"):
            print(f"  {prop['description']}")  # description is help text
        hints = [
            "options: " + "/".join(prop["enum"]) if prop.get("enum") else str(prop.get("type")),
        ]
        if "default" in prop:
            hints.append(f"default: {prop['default']}")
        while True:
            raw = input(f"  {label} ({', '.join(hints)}): ").strip()
            if raw == "!decline":
                return ElicitResult(action="decline")
            if raw == "!cancel":
                return ElicitResult(action="cancel")
            if raw == "":
                if "default" in prop:
                    content[key] = prop["default"]
                elif key in required:
                    print(f'  "{label}" is required.')
                    continue
                break
            ok, value = _parse_input(prop, raw)
            error = validate_field(key, prop, value) if ok else f'"{label}" is not a valid {prop.get("type")}'
            if error is not None:
                print(f"  {error}")
                continue
            content[key] = value
            break
    if errors := validate_content(params, content):
        # Should not happen after per-field checks, but never send invalid data.
        print(f"  Input failed validation: {'; '.join(errors)}")
        return ElicitResult(action="cancel")
    return ElicitResult(action="accept", content=content)


# ---------------------------------------------------------------------------
# The handler itself. `handle_elicitation` is what the native driver's
# `elicitation_callback` runs; `handle_input_request` adapts a bare embedded
# request from an `input_requests` map (an `ElicitRequest` with
# method/params) for the manual loop.
# ---------------------------------------------------------------------------


async def handle_elicitation(params: ElicitRequestParams) -> ElicitResult:
    if isinstance(params, ElicitRequestURLParams):
        # URL mode: display the full URL, never pre-fetch, never auto-navigate.
        # This console host only reports consent; a real host opens the system
        # browser after explicit user approval.
        print(f"  [url-mode] {params.message}")
        print(f"  [url-mode] target: {params.url} — open it in your browser, then continue.")
        return ElicitResult(action="accept")

    print(f"  elicitation: {params.message}")

    for policy in POLICIES:
        if (verdict := policy(params)) is not None:
            return verdict  # answered or denied by policy

    if _scripted_answers is not None:
        return next_scripted_answer(params)
    # input() blocks; keep the event loop responsive.
    return await anyio.to_thread.run_sync(_render_form_sync, params)


async def handle_input_request(request: InputRequest) -> ElicitResult:
    if request.method != "elicitation/create":
        # This host supports elicitation only (sampling is deprecated; roots n/a).
        raise RuntimeError(f"Unsupported input request: {request.method}")
    return await handle_elicitation(request.params)

File: chapter06/mrtr-client/typescript/src/input-handler.ts
// handleInputRequest: the client host's side of an elicitation.
//
// Two concerns from Chapter 6, Section 6.2.5:
//   1. Form rendering — a schema-driven console form with labels, help text,
//      defaults, the accept/decline/cancel three-action model, and validation
//      of the user's input against the requested schema before responding.
//   2. Policy hooks — pre-user policies that can auto-answer or deny a
//      request without it ever reaching the user.
//
// Modes:
//   - Interactive (default): renders the form on the terminal.
//   - Scripted: MRTR_ANSWERS holds a JSON array of predetermined answers,
//     consumed in order — one per elicitation. An entry is either a content
//     object (meaning accept) or { "action": "decline" | "cancel" }.
//     Example: MRTR_ANSWERS='[{"title":"Sync","duration":"30"},{"confirm":true}]'
//
// MRTR_POLICY=autoconfirm enables the demo auto-answer policy.

import * as readline from 'node:readline/promises';
import type { ElicitResult } from '@modelcontextprotocol/client';

type SchemaProperty = {
  type?: string;
  title?: string;
  description?: string;
  enum?: string[];
  default?: unknown;
  minLength?: number;
  maxLength?: number;
  minimum?: number;
  maximum?: number;
  format?: string;
};

type FormParams = {
  message: string;
  requestedSchema: {
    type: 'object';
    properties: Record<string, SchemaProperty>;
    required?: string[];
  };
};

// ---------------------------------------------------------------------------
// Policy hooks: run before the user ever sees the request. A policy returns a
// full response to short-circuit the form, or undefined to pass.
// ---------------------------------------------------------------------------

type Policy = (params: FormParams) => ElicitResult | undefined;

/** Form mode must never collect credentials; deny requests that look like it. */
const denyCredentialSolicitation: Policy = (params) => {
  const suspicious = /password|api[ -]?key|token|secret|credential/i;
  for (const [key, prop] of Object.entries(params.requestedSchema.properties)) {
    const texts = [key, prop.title ?? '', prop.description ?? ''].join(' ');
    if (prop.format === 'password' || suspicious.test(texts)) {
      console.log(`  [policy] declined: field "${key}" appears to solicit credentials`);
      return { action: 'decline' };
    }
  }
  return undefined;
};

/** Demo auto-answer policy: accept pure-confirmation forms (all-boolean) with their defaults. */
const autoConfirm: Policy = (params) => {
  if (process.env['MRTR_POLICY'] !== 'autoconfirm') return undefined;
  const props = Object.entries(params.requestedSchema.properties);
  if (props.length === 0 || !props.every(([, p]) => p.type === 'boolean')) return undefined;
  const content: Record<string, boolean> = {};
  for (const [key, prop] of props) {
    content[key] = typeof prop.default === 'boolean' ? prop.default : true;
  }
  console.log(`  [policy] auto-answered confirmation form: ${JSON.stringify(content)}`);
  return { action: 'accept', content };
};

const policies: Policy[] = [denyCredentialSolicitation, autoConfirm];

// ---------------------------------------------------------------------------
// Validation: check content against the requested schema before responding —
// a correctness concern (the server expects this shape) and a security one.
// ---------------------------------------------------------------------------

function validateField(key: string, prop: SchemaProperty, value: unknown): string | undefined {
  if (prop.enum !== undefined) {
    if (typeof value !== 'string' || !prop.enum.includes(value)) {
      return `"${key}" must be one of: ${prop.enum.join(', ')}`;
    }
    return undefined;
  }
  switch (prop.type) {
    case 'string':
      if (typeof value !== 'string') return `"${key}" must be a string`;
      if (prop.minLength !== undefined && value.length < prop.minLength)
        return `"${key}" must be at least ${prop.minLength} characters`;
      if (prop.maxLength !== undefined && value.length > prop.maxLength)
        return `"${key}" must be at most ${prop.maxLength} characters`;
      return undefined;
    case 'number':
    case 'integer':
      if (typeof value !== 'number' || Number.isNaN(value)) return `"${key}" must be a number`;
      if (prop.type === 'integer' && !Number.isInteger(value)) return `"${key}" must be an integer`;
      if (prop.minimum !== undefined && value < prop.minimum)
        return `"${key}" must be >= ${prop.minimum}`;
      if (prop.maximum !== undefined && value > prop.maximum)
        return `"${key}" must be <= ${prop.maximum}`;
      return undefined;
    case 'boolean':
      return typeof value === 'boolean' ? undefined : `"${key}" must be a boolean`;
    default:
      return `"${key}" has an unsupported schema type: ${prop.type}`;
  }
}

function validateContent(params: FormParams, content: Record<string, unknown>): string[] {
  const errors: string[] = [];
  const { properties, required = [] } = params.requestedSchema;
  for (const key of required) {
    if (content[key] === undefined) errors.push(`"${key}" is required`);
  }
  for (const [key, value] of Object.entries(content)) {
    const prop = properties[key];
    if (prop === undefined) {
      errors.push(`"${key}" is not in the requested schema`);
      continue;
    }
    const error = validateField(key, prop, value);
    if (error !== undefined) errors.push(error);
  }
  return errors;
}

/** Pre-populate defaults, per the schema-driven rendering rules. */
function defaultsOf(params: FormParams): Record<string, unknown> {
  const content: Record<string, unknown> = {};
  for (const [key, prop] of Object.entries(params.requestedSchema.properties)) {
    if (prop.default !== undefined) content[key] = prop.default;
  }
  return content;
}

// ---------------------------------------------------------------------------
// Scripted answers (for smoke tests / headless runs).
// ---------------------------------------------------------------------------

const scriptedAnswers: unknown[] | undefined = process.env['MRTR_ANSWERS']
  ? (JSON.parse(process.env['MRTR_ANSWERS']) as unknown[])
  : undefined;

function nextScriptedAnswer(params: FormParams): ElicitResult {
  const entry = scriptedAnswers!.shift();
  if (entry === undefined) throw new Error('MRTR_ANSWERS ran out of scripted answers');
  const record = entry as Record<string, unknown>;
  if (record['action'] === 'decline' || record['action'] === 'cancel') {
    console.log(`  [scripted] ${record['action']}`);
    return { action: record['action'] as 'decline' | 'cancel' };
  }
  const content = { ...defaultsOf(params), ...record };
  const errors = validateContent(params, content);
  if (errors.length > 0) throw new Error(`Scripted answer failed validation: ${errors.join('; ')}`);
  console.log(`  [scripted] accept: ${JSON.stringify(content)}`);
  return { action: 'accept', content: content as NonNullable<ElicitResult['content']> };
}

// ---------------------------------------------------------------------------
// Interactive form rendering.
// ---------------------------------------------------------------------------

function parseInput(prop: SchemaProperty, raw: string): { ok: true; value: unknown } | { ok: false } {
  if (prop.enum !== undefined || prop.type === 'string') return { ok: true, value: raw };
  if (prop.type === 'number' || prop.type === 'integer') {
    const n = Number(raw);
    return Number.isNaN(n) ? { ok: false } : { ok: true, value: n };
  }
  if (prop.type === 'boolean') {
    if (/^(y|yes|true)$/i.test(raw)) return { ok: true, value: true };
    if (/^(n|no|false)$/i.test(raw)) return { ok: true, value: false };
    return { ok: false };
  }
  return { ok: false };
}

async function renderForm(params: FormParams): Promise<ElicitResult> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  try {
    console.log(`\n${params.message}`);
    console.log('(enter a value, or !decline / !cancel; blank accepts the default)');
    const content: Record<string, unknown> = {};
    const required = params.requestedSchema.required ?? [];
    for (const [key, prop] of Object.entries(params.requestedSchema.properties)) {
      const label = prop.title ?? key; // title is the display label
      if (prop.description) console.log(`  ${prop.description}`); // description is help text
      const hints = [
        prop.enum ? `options: ${prop.enum.join('/')}` : prop.type,
        prop.default !== undefined ? `default: ${prop.default}` : undefined,
      ].filter(Boolean);
      for (;;) {
        const raw = (await rl.question(`  ${label} (${hints.join(', ')}): `)).trim();
        if (raw === '!decline') return { action: 'decline' };
        if (raw === '!cancel') return { action: 'cancel' };
        if (raw === '') {
          if (prop.default !== undefined) content[key] = prop.default;
          else if (required.includes(key)) {
            console.log(`  "${label}" is required.`);
            continue;
          }
          break;
        }
        const parsed = parseInput(prop, raw);
        const error = parsed.ok ? validateField(key, prop, parsed.value) : `"${label}" is not a valid ${prop.type}`;
        if (error !== undefined) {
          console.log(`  ${error}`);
          continue;
        }
        content[key] = (parsed as { value: unknown }).value;
        break;
      }
    }
    const errors = validateContent(params, content);
    if (errors.length > 0) {
      // Should not happen after per-field checks, but never send invalid data.
      console.log(`  Input failed validation: ${errors.join('; ')}`);
      return { action: 'cancel' };
    }
    return { action: 'accept', content: content as NonNullable<ElicitResult['content']> };
  } finally {
    rl.close();
  }
}

// ---------------------------------------------------------------------------
// The handler itself. Accepts either a bare embedded request from an
// `inputRequests` map or the JSON-RPC request the SDK's native driver
// dispatches — both carry { method, params }.
// ---------------------------------------------------------------------------

export async function handleInputRequest(request: unknown): Promise<ElicitResult> {
  const { method, params } = request as { method?: string; params?: Record<string, unknown> };
  if (method !== 'elicitation/create' || params === undefined) {
    // This host supports elicitation only (sampling is deprecated; roots n/a).
    throw new Error(`Unsupported input request: ${method}`);
  }

  if ((params['mode'] ?? 'form') === 'url') {
    // URL mode: display the full URL, never pre-fetch, never auto-navigate.
    // This console host only reports consent; a real host opens the system
    // browser after explicit user approval.
    console.log(`  [url-mode] ${params['message']}`);
    console.log(`  [url-mode] target: ${params['url']} — open it in your browser, then continue.`);
    return { action: 'accept' };
  }

  const form = params as unknown as FormParams;
  console.log(`  elicitation: ${form.message}`);

  for (const policy of policies) {
    const verdict = policy(form);
    if (verdict !== undefined) return verdict; // answered or denied by policy
  }

  if (scriptedAnswers !== undefined) return nextScriptedAnswer(form);
  return renderForm(form);
}
